### ISSUE TITLE
Colony member attach positions become slightly wrong when far away from the lead cell

### ISSUE THREAD

**Issue by hhyyrylainen** [Revolutionary-Games/Thrive#4666]
When binding through 3-4 cells the cell positions start to get really off sometimes and cause problems:
![2023-12-14_16 26 51 6719](https://github.com/Revolutionary-Games/Thrive/assets/3796411/1fe237c5-cded-46ce-aaac-cf75bd4efb0d)

The issue is most likely in `MicrobeColonyHelpers.CalculateColonyMemberAttachPosition` where I put a TODO on the line I suspect is the problem.

The problem *might* be still in `GetNewRelativeTransform` though when recursively adding to colony members but I doubt that at least a bit.

**Comment by Kemikal1** [comment]
I kinda graduated from doing the math Godot already does for you. I also had this problem in my projects.
I have a REVOLUTIONARY solution to this:
  1.Save the global (rotation + translation) of the thing you are trying to parent(cell in this case).
  2.Add the cell as a child to parent.
  3.Reset the property things with the ones you saved in step 1.

**Comment by hhyyrylainen** [comment]
That could work. However with the ECS refactor, the gameplay logic no longer depends on the Godot nodes so directly using such an approach is not possible (as the internal game simulation doesn't use Godot at all and Godot nodes are only used to display the results afterwards). But looking at the math for how Godot preserves the world location and rotation of objects when parenting, probably has the answer in it. The hard part is double checking (or rewriting from scratch) the current math used in the game to calculate the positions and correct the math error that seems to start building up as things get farther from the lead cell.

**Comment by Kemikal1** [comment]
F
So now you need to calculated the transforms everytime a new child is paranted?
I dont about the architecture that much but, my guess would be to maybe just dont modify the world position and rotation of the cell and leave as is.

**Comment by hhyyrylainen** [comment]
The previous code did a similar calculation (and it also resulted in similarly wrong attach positions occasionally). It's simply required to calculate the offset from the new attached cell to the original, because otherwise it is not possible to make it look like the new cell is moves along with the parent. Also the position and rotation is used by the physics system to correctly position the physics subshape of the cell (this makes sure it is visually and physically in the same position).

**Comment by Kemikal1** [comment]
I know, i know  about why the code is and was like this. But this kind of code will be needed every time you reparent a node to another and keep the childs position while rotating. Which is why i said that now youll need to do similar calculations for everything that involves reparenting. I also think that you should use transforms instead of quats cause that will unify the offsets and rotation into one.

**Comment by hhyyrylainen** [comment]
The physics engine takes in data as position (Vec3) and rotation (Quat), converting between transforms and that data would add a lot of extra data conversions. Right now only one conversion is needed from the physics engine provided position and rotation + the scale property from visual instance component to be combined into a single transform and sent to Godot for renderable positioning. Also physics really hates scale, so it is also a bug prevention measure that the data format doesn't support that at all. 

In the new system parenting is defined with a component like:
```
    public struct AttachedToEntity
    {
        /// <summary>
        ///   Entity this is attached to. Should be valid whenever this component exists
        /// </summary>
        public Entity AttachedTo;

        /// <summary>
        ///   Position relative to the parent entity
        /// </summary>
        public Vector3 RelativePosition;

        /// <summary>
        ///   Rotation relative to the parent entity
        /// </summary>
        public Quat RelativeRotation;
    }
```

As far as I've seen, I got the parented position math exactly right (I've not seen any nor has anyone else any bugs). The only part where there is a bug is calculating the necessary relative rotation and position when attaching cells to a colony.

**Comment by Kemikal1** [comment]
Then what ive said in my first com SHOULD still apply.

**Comment by hhyyrylainen** [comment]
It does and so does my reply. The math behind the parenting operation is what is needed.

The colony attach operation moves the new cell so that it is touching membranes with an existing cell. This was the case before even the ECS refactor, and this is the source where the custom attach positioning logic originated from. I just ported that over to the ECS system and managed to fix some cases, but there is still [this issue](https://github.com/Revolutionary-Games/Thrive/issues/4666) with the math.

**Comment by Kemikal1** [comment]
I dont think and i dont remember there was a need ever to  move the new cell, just have its relative position to master in its place updated.
You jsut said you have the math for the parenting done right so there shouldnt be a need anymore to recaulculate it for colony cells up the family tree (or down however you imagine it).

**Comment by hhyyrylainen** [comment]
As far as I remember, even the first colony binding code had the logic for making sure the membrane points are snapped together, because otherwise the cells might not be actually touching after binding happens.

I'd have to re-check the code, but it could be easier to do the following:
- Calculate final world position the cell should be attached at
- Calculate the full offset from the colony lead cell to the world position (note that this is not the same math as resolving the parented object's final world position based on the parent and offset, rather this is the reverse but due to vector math the reverse calculation needs thought put into it to ensure it is done correctly)

Rather than how it is currently:
- Calculate offset to the currently hit cell
- Combine offset of the hit cell with the calculated offset (I've tried various orders of multiplying all the stuff together but it goes wrong for some reason)
- Use the combined offset to parent the new colony member to the colony lead cell

**Comment by hhyyrylainen** [comment]
Based on the world position approach I realized I had made at least one serious logic issue with the previous code where I had mixed up the colony lead cell position info and the actually collided cell. Thinking through with the world position based approach this was much easier to make a fix. I did make one bug which I was able to fix by implementing a visual debugging line system. The problem was that I forgot in one place to multiply a position by a reverse rotation.

Here's the changes: https://github.com/Revolutionary-Games/Thrive/pull/4881
